Fixed #18620 -- Made ContentTypes shortcut view prefer current site if available.

czpython opened the pull request:
https://code.djangoproject.com/ticket/18620

Initial patch: https://github.com/django/django/pull/203/

I had to adapt an existing test because the proposed solution will always use an absolute url if one is available. The previous solution would always use one only if the sites framework was not installed.

czpython commented:
Hello @carltongibson,
Thanks for the follow-up.
The code block in question was there prior to this change, I'll gladly check if there's tests for it and can push some in this same pr.

carltongibson commented:
Hi @czpython. 

Yes it was. There's no explicit test covering it, but previously it was at least exercised but the test suite. It's now not run at all. 

Just from a "black box" perspective I think it's worth adding a test case here for this second branch. 

Thanks!

carltongibson commented:
Hi @czpython. I've rebased this and added a test case for the _Other Site_ case. (Please take a look and make sure you're happy.) 

Assuming the build passed I will mark this Ready for checkin. Thanks for input!

timgraham commented on tests/contenttypes_tests/test_views.py:
```diff
@@ -104,7 +104,55 @@ def test_shortcut_view_with_null_site_fk(self, get_model):
+        """
+        get_model.side_effect = lambda *args, **kwargs: MockSite if args[0] == 'sites.Site' else ModelWithM2MToSite
+
+        # get_current_site will lookup a Site object,
```
Wrap lines closer to 79 characters and use () when referring to a function.
```
# get_current_site() will lookup a Site object, so these must match the
# domains in the MockSite model.
```

timgraham commented on tests/contenttypes_tests/test_views.py:
```diff
@@ -104,7 +104,55 @@ def test_shortcut_view_with_null_site_fk(self, get_model):
+            Site(pk=2, domain='example-2.com', name='example-2.com'),
+            Site(pk=3, domain='example-3.com', name='example-3.com'),
+        ])
+
```
Less blank lines please. Some comments might take the place of blank lines. I'm not sure if the docstring does a good job at describing the purpose of the entire test. About some code, Carlton mentioned, "There's no explicit test covering it, but previously it was at least exercised but the test suite. It's now not run at all." -- is part of the test passing even before the fix? If so, we'd prefer to add that as a separate commit that precedes the fix to make it more clear what issue is fixed here and so that if the fix is reverted, for example, then that additional test coverage can stay.

timgraham commented on django/contrib/contenttypes/views.py:
```diff
@@ -43,27 +43,37 @@ def shortcut(request, content_type_id, object_id):
-                    pass
-                if object_domain is not None:
+                site_qs = getattr(obj, field.name).all()
+
```
Chop blank line.

timgraham commented on django/contrib/contenttypes/views.py:
```diff
@@ -43,27 +43,37 @@ def shortcut(request, content_type_id, object_id):
+                # Caveat: In the case of multiple related Sites, this just
+                # selects the *first* one, which is arbitrary.
+                site = site_qs.first()
+
```
Chop blank line.

timgraham commented on django/contrib/contenttypes/views.py:
```diff
@@ -43,27 +43,37 @@ def shortcut(request, content_type_id, object_id):
     if apps.is_installed('django.contrib.sites'):
         Site = apps.get_model('sites.Site')
+        check_many_to_one = True
 
```
I'd chop this blank line.

timgraham commented on django/contrib/contenttypes/views.py:
```diff
@@ -43,27 +43,37 @@ def shortcut(request, content_type_id, object_id):
+
         # Next, look for a many-to-one relationship to Site.
-        if object_domain is None:
+        if check_many_to_one:
```
I think this could be replaced with for/else and adding `break` where `check_many_to_one = False` is.

timgraham commented on tests/contenttypes_tests/test_views.py:
```diff
@@ -104,7 +104,55 @@ def test_shortcut_view_with_null_site_fk(self, get_model):
+        # get_current_site will lookup a Site object,
+        # so these need to match the same domains in the MockSite model.
+        Site.objects.bulk_create([
+            Site(pk=2, domain='example-2.com', name='example-2.com'),
```
Is there a need to hardcode pks? This is generally to be avoided, I think.

timgraham commented on tests/contenttypes_tests/test_views.py:
```diff
@@ -104,7 +104,55 @@ def test_shortcut_view_with_null_site_fk(self, get_model):
+        # so these need to match the same domains in the MockSite model.
+        Site.objects.bulk_create([
+            Site(pk=2, domain='example-2.com', name='example-2.com'),
+            Site(pk=3, domain='example-3.com', name='example-3.com'),
```
For easier typing and consistency with elsewhere, I'd omit the dash in the domains and names.

timgraham commented on tests/contenttypes_tests/test_views.py:
```diff
@@ -104,7 +104,55 @@ def test_shortcut_view_with_null_site_fk(self, get_model):
+        other_obj = ModelWithM2MToSite.objects.create(title='Not Linked to Current Site')
+        other_obj.sites.add(MockSite.objects.get(pk=3))
+
+        shortcut_url = '/shortcut/%s/%s/' % (ct.pk, other_obj.pk)
```
I'd omit the `shortcut_url` variable and put this directly in the `get()`.

czpython commented:
Thanks for the review @timgraham @carltongibson.
I agree that we should split the commits.
I'll address the PR comments over the weekend.

czpython commented:
@carltongibson @timgraham 
I've rebased with latest master and added two separate commits.
The first one is the missing test and the second is the new change in behavior with a change in the test of the first commit.

timgraham commented:
There's a test failure on databases besides SQLite. Is there any way to break up the tests into smaller methods? Maybe a separate class with `setUpTestData()` would be appropriate?

carltongibson commented:
Unless the middle commit can go first by itself that is. (I didn't look at that.)

czpython commented:
I think it can, will check.

czpython commented:
@carltongibson 
Pushed the rebase.
There's a new failing test even though nothing changed, I don't think is related to this change.
Maybe that one build can be retried?
